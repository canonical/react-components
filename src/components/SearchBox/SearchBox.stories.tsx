import React from "react";
import { Meta, StoryObj } from "@storybook/react";

import SearchBox from "./SearchBox";

const meta: Meta<typeof SearchBox> = {
  component: SearchBox,
  tags: ["autodocs"],
};

export default meta;

type Story = StoryObj<typeof SearchBox>;

export const Default: Story = {
  name: "Default",
};

export const Disabled: Story = {
  render: () => <SearchBox disabled />,
  name: "Disabled",
};

/**
 * If you wish to control the value of the input via external state you can set
the `externallyControlled` prop and provide an `onChange` method to update the
state and the `value` from state.
 */
export const ExternalState: Story = {
  render: () => (
    <SearchBox
      externallyControlled
      shouldRefocusAfterReset
      onChange={() => {}}
      value="..."
    />
  ),

  name: "External state",
};

export const Navigation: Story = {
  render: () => (
    <header id="navigation" className="p-navigation">
      <div className="p-navigation__row--full-width">
        <div className="p-navigation__banner">
          <div className="p-navigation__logo">
            <a className="p-navigation__item" href="#test">
              <img
                className="p-navigation__image"
                src="https://assets.ubuntu.com/v1/5d6da5c4-logo-canonical-aubergine.svg"
                alt=""
                width="95"
              />
            </a>
          </div>
          <a
            href="#navigation"
            className="p-navigation__toggle--open"
            title="menu"
          >
            Menu
          </a>
          <a
            href="#navigation-closed"
            className="p-navigation__toggle--close"
            title="close menu"
          >
            Close menu
          </a>
        </div>
        <nav className="p-navigation__nav">
          <span className="u-off-screen">
            <a href="#main-content">Jump to main content</a>
          </span>
          <ul className="p-navigation__items" role="menu">
            <li className="p-navigation__item" role="menuitem">
              <a href="#test" className="p-navigation__link">
                Products
              </a>
            </li>
            <li className="p-navigation__item" role="menuitem">
              <a href="#test" className="p-navigation__link">
                Services
              </a>
            </li>
            <li className="p-navigation__item" role="menuitem">
              <a href="#test" className="p-navigation__link">
                Partners
              </a>
            </li>
            <li className="p-navigation__item" role="menuitem">
              <a href="#test" className="p-navigation__link">
                About
              </a>
            </li>
            <li className="p-navigation__item" role="menuitem">
              <a href="#test" className="p-navigation__link">
                Partners
              </a>
            </li>
          </ul>
          <SearchBox />
        </nav>
      </div>
    </header>
  ),

  name: "Navigation",
};
