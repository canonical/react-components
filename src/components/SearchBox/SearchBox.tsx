import classNames from "classnames";
import React, { HTMLProps, KeyboardEvent, useRef } from "react";

import Icon from "../Icon";

import type { ClassName, PropsWithSpread } from "types";

export enum Label {
  Clear = "Clear search field",
  Search = "Search",
}

export type Props = PropsWithSpread<
  {
    /**
     * Whether autocomplete should be enabled for the search input.
     */
    autocomplete?: "on" | "off";
    /**
     * Optional classes to pass to the form element.
     */
    className?: ClassName;
    /**
     * Whether the input and buttons should be disabled.
     */
    disabled?: boolean;
    /**
     * Whether the input value will be controlled via external state.
     */
    externallyControlled?: boolean;
    /**
     * A function that will be called when the input value changes.
     */
    onChange?: (inputValue: string) => void;
    /**
     * A function that is called when the user clicks the search icon or presses enter
     */
    onSearch?: () => void;
    /**
     * A function that is called when the user clicks the reset icon
     */
    onClear?: () => void;
    /**
     * A search input placeholder message.
     */
    placeholder?: string;
    /**
     * Whether the search input should receive focus after pressing the reset button
     */
    shouldRefocusAfterReset?: boolean;
    /**
     * A ref that is passed to the input element.
     */
    ref?: string;
    /**
     * The value of the search input when the state is externally controlled.
     */
    value?: string;
  },
  HTMLProps<HTMLInputElement>
>;

/**
 * This is a [React](https://reactjs.org/) component for the Vanilla [SearchBox](https://docs.vanillaframework.io/patterns/search-box/).
 *
 * Search boxes enable search functionality on a page and are typically used in a navigation bar. SearchBox elements should be wrapped within a form element that handles its own submit action.
 */
const SearchBox = React.forwardRef<HTMLInputElement, Props>(
  (
    {
      autocomplete = "on",
      className,
      disabled,
      externallyControlled,
      onChange,
      onSearch,
      onClear,
      placeholder = "Search",
      shouldRefocusAfterReset,
      value,
      ...props
    }: Props,
    forwardedRef
  ): JSX.Element => {
    const internalRef = useRef<HTMLInputElement>();
    const resetInput = () => {
      onChange?.("");
      onClear?.();
      if (internalRef.current) {
        internalRef.current.value = "";
        if (shouldRefocusAfterReset) internalRef.current.focus();
      }
    };

    const triggerSearch = () => {
      onSearch && onSearch();
    };

    const onKeyDown = (e: KeyboardEvent<HTMLInputElement>) => {
      if (e.key === "Enter" && internalRef.current.checkValidity()) {
        internalRef.current.blur();
        triggerSearch();
      }
    };

    return (
      <div className={classNames("p-search-box", className)}>
        <label className="u-off-screen" htmlFor="search">
          {placeholder || "Search"}
        </label>
        <input
          autoComplete={autocomplete}
          className="p-search-box__input"
          disabled={disabled}
          id="search"
          name="search"
          onChange={(evt) => onChange?.(evt.target.value)}
          onKeyDown={onKeyDown}
          placeholder={placeholder}
          ref={(input) => {
            internalRef.current = input;
            // Handle both function and object refs.
            if (typeof forwardedRef === "function") {
              forwardedRef(input);
            } else if (forwardedRef) {
              forwardedRef.current = input;
            }
          }}
          type="search"
          defaultValue={externallyControlled ? undefined : value}
          value={externallyControlled ? value : undefined}
          {...props}
        />
        {value && (
          <button
            className="p-search-box__reset"
            disabled={disabled}
            onClick={resetInput}
            type="reset"
          >
            <Icon name="close">{Label.Clear}</Icon>
          </button>
        )}
        <button
          className="p-search-box__button"
          disabled={disabled}
          onClick={triggerSearch}
        >
          <Icon name="search">{Label.Search}</Icon>
        </button>
      </div>
    );
  }
);

SearchBox.displayName = "SearchBox";

export default SearchBox;
