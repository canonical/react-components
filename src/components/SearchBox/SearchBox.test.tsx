import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import React from "react";

import SearchBox, { Label } from "./SearchBox";

describe("SearchBox ", () => {
  it("renders", () => {
    render(<SearchBox />);
    // Get the wrapping element for the snapshot.
    // eslint-disable-next-line testing-library/no-node-access
    expect(document.querySelector(".p-search-box")).toMatchSnapshot();
  });

  it("shows the clear button when there is a value", () => {
    render(<SearchBox value="admin" />);
    expect(
      screen.getByRole("button", { name: Label.Clear })
    ).toBeInTheDocument();
  });

  it("can externally control the value", () => {
    const { rerender } = render(
      <SearchBox externallyControlled onChange={jest.fn()} value="admin" />
    );
    expect(screen.getByRole("searchbox")).toHaveAttribute("value", "admin");
    rerender(
      <SearchBox externallyControlled onChange={jest.fn()} value="new-admin" />
    );
    expect(screen.getByRole("searchbox")).toHaveAttribute("value", "new-admin");
  });

  it("should call onSearch prop", async () => {
    const onSearchMock = jest.fn();
    render(<SearchBox onSearch={onSearchMock} />);
    await userEvent.click(screen.getByRole("button", { name: Label.Search }));
    expect(onSearchMock).toBeCalled();
  });

  it("should call onChange prop", async () => {
    const onChangeMock = jest.fn();
    render(<SearchBox onChange={onChangeMock} />);
    await userEvent.type(screen.getByRole("searchbox"), "test");
    expect(onChangeMock).toBeCalledWith("test");
  });

  it("renders extra props", () => {
    render(<SearchBox data-testid="testID" />);
    expect(screen.getByRole("searchbox")).toHaveAttribute(
      "data-testid",
      "testID"
    );
  });

  it("accepts a ref for the input", () => {
    const ref = React.createRef<HTMLInputElement>();
    render(<SearchBox ref={ref} />);
    ref.current?.focus();
    expect(screen.getByRole("searchbox")).toHaveFocus();
  });
});
