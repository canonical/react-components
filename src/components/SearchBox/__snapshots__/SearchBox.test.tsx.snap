// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBox  renders 1`] = `
<div
  class="p-search-box"
>
  <label
    class="u-off-screen"
    for="search"
  >
    Search
  </label>
  <input
    autocomplete="on"
    class="p-search-box__input"
    id="search"
    name="search"
    placeholder="Search"
    type="search"
    value=""
  />
  <button
    class="p-search-box__button"
  >
    <i
      class="p-icon--search"
    >
      Search
    </i>
  </button>
</div>
`;
